### ISSUE TITLE
PacBio HiFi HG002 chr20 results

### ISSUE THREAD

**Issue by tahashmi** [kishwarshafin/pepper#42]
Hi Kishwar,

I have just tested [PEPPER](https://github.com/kishwarshafin/pepper/blob/r0.4/docs/pipeline_singularity/HiFi_variant_calling_singularity.md) PacBio HiFi variant calling workflow. My results are bit different for both cases, 
1. With your given chr20 input data, I get: 

```
[W] overlapping records at chr6:29747433 for sample 0
[W] Variants that overlap on the reference allele: 5
[I] Total VCF records:         4048342
[I] Non-reference VCF records: 4048342
[W] overlapping records at chr20:3057231 for sample 0
[W] Variants that overlap on the reference allele: 3
[I] Total VCF records:         213510
[I] Non-reference VCF records: 124031
2021-03-09 05:59:36,413 WARNING  Creating template for vcfeval. You can speed this up by supplying a SDF template that corresponds to /home/tahmad/tahmad/testing/pacbio_dv_pepper/hifi-case-study/input/data/GRCh38_no_alt.chr20.fa
/home/tahmad/.local/lib/python2.7/site-packages/pandas/core/computation/check.py:19: UserWarning: The installed version of numexpr 2.4.3 is not supported in pandas and will be not be used
The minimum supported version is 2.6.1

  ver=ver, min_ver=_MIN_NUMEXPR_VERSION), UserWarning)
Benchmarking Summary:
  Type Filter  TRUTH.TOTAL  TRUTH.TP  TRUTH.FN  QUERY.TOTAL  QUERY.FP  QUERY.UNK  FP.gt  METRIC.Recall  METRIC.Precision  METRIC.Frac_NA  METRIC.F1_Score  TRUTH.TOTAL.TiTv_ratio  QUERY.TOTAL.TiTv_ratio  TRUTH.TOTAL.het_hom_ratio  QUERY.TOTAL.het_hom_ratio
 INDEL    ALL        11256     11172        84        22764        96      11083     52       0.992537          0.991782        0.486865         0.992159                     NaN                     NaN                   1.561710                   2.223132
 INDEL   PASS        11256     11172        84        22764        96      11083     52       0.992537          0.991782        0.486865         0.992159                     NaN                     NaN                   1.561710                   2.223132
   SNP    ALL        71333     71278        55       102480         9      31114      7       0.999229          0.999874        0.303610         0.999551                2.314904                1.879008                   1.715978                   1.937048
   SNP   PASS        71333     71278        55       102480         9      31114      7       0.999229          0.999874        0.303610         0.999551                2.314904                1.879008                   1.715978                   1.937048

```
2. When aligning precisionFDA HG002 FASTQ through minimap2 for chr20 (command: `minimap2', '-a',  '-k 19',  '-O 5,56',  '-E 4,1',  '-B 5',  '-z 400,50',  '-r 2k',  '--eqx',  '--secondary=no', '-t 32'`), I get:
```
[W] overlapping records at chr6:29747433 for sample 0
[W] Variants that overlap on the reference allele: 5
[I] Total VCF records:         4048342
[I] Non-reference VCF records: 4048342
[W] overlapping records at chr20:16032761 for sample 0
[W] Variants that overlap on the reference allele: 3
[I] Total VCF records:         197238
[I] Non-reference VCF records: 117736
2021-03-24 21:59:36,438 WARNING  Creating template for vcfeval. You can speed this up by supplying a SDF template that corresponds to /home/tahmad/tahmad/testing/pacbio_dv_pepper/hifi-case-study/input/data/GRCh38_no_alt.chr20.fa
Benchmarking Summary:
  Type Filter  TRUTH.TOTAL  TRUTH.TP  TRUTH.FN  QUERY.TOTAL  QUERY.FP  QUERY.UNK  FP.gt  FP.al  METRIC.Recall  METRIC.Precision  METRIC.Frac_NA  METRIC.F1_Score  TRUTH.TOTAL.TiTv_ratio  QUERY.TOTAL.TiTv_ratio  TRUTH.TOTAL.het_hom_ratio  QUERY.TOTAL.het_hom_ratio
 INDEL    ALL        11256     11075       181        22119       191      10453    105     41       0.983920          0.983628        0.472580         0.983774                     NaN                     NaN                   1.561710                   2.158347
 INDEL   PASS        11256     11075       181        22119       191      10453    105     41       0.983920          0.983628        0.472580         0.983774                     NaN                     NaN                   1.561710                   2.158347
   SNP    ALL        71333     71276        57        96834         9      25470      7      1       0.999201          0.999874        0.263027         0.999537                2.314904                1.991077                   1.715978                   2.018434
   SNP   PASS        71333     71276        57        96834         9      25470      7      1       0.999201          0.999874        0.263027         0.999537                2.314904                1.991077                   1.715978                   2.018434

```
I am wondering if something is being missed from my side to get 100% same results as you have mentioned? Thanks.

**Comment by kishwarshafin** [comment]
@tahashmi , 
I just re-ran the PacBio-HiFi use-case on a single node on GCP. I was able to reproduce the numbers exactly how we reported:
```bash
 Type Filter  TRUTH.TOTAL  TRUTH.TP  TRUTH.FN  QUERY.TOTAL  QUERY.FP  QUERY.UNK  FP.gt  FP.al  METRIC.Recall  METRIC.Precision  METRIC.Frac_NA  METRIC.F1_Score  TRUTH.TOTAL.TiTv_ratio  QUERY.TOTAL.TiTv_ratio  TRUTH.TOTAL.het_hom_ratio  QUERY.TOTAL.het_hom_ratio
INDEL    ALL        11256     11175        81        22787        94      11105     52     33       0.992804          0.991953        0.487339         0.992378                     NaN                     NaN                   1.561710                   2.222758
INDEL   PASS        11256     11175        81        22787        94      11105     52     33       0.992804          0.991953        0.487339         0.992378                     NaN                     NaN                   1.561710                   2.222758
  SNP    ALL        71333     71277        56       102592         8      31226      7      1       0.999215          0.999888        0.304371         0.999551                2.314904                1.878355                   1.715978                   1.936050
  SNP   PASS        71333     71277        56       102592         8      31226      7      1       0.999215          0.999888        0.304371         0.999551                2.314904                1.878355                   1.715978                   1.936050
```

My setup:
Single node: n1-standard-96 (96 vCPUs, 360 GB memory)
CPUs: 
```bash
Architecture:        x86_64
CPU op-mode(s):      32-bit, 64-bit
Byte Order:          Little Endian
CPU(s):              96
On-line CPU(s) list: 0-95
Thread(s) per core:  2
Core(s) per socket:  24
Socket(s):           2
NUMA node(s):        2
Vendor ID:           GenuineIntel
CPU family:          6
Model:               85
Model name:          Intel(R) Xeon(R) CPU @ 2.00GHz
Stepping:            3
CPU MHz:             2000.182
BogoMIPS:            4000.36
Hypervisor vendor:   KVM
Virtualization type: full
L1d cache:           32K
L1i cache:           32K
L2 cache:            1024K
L3 cache:            39424K
NUMA node0 CPU(s):   0-23,48-71
NUMA node1 CPU(s):   24-47,72-95
Flags:               fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology nonstop_tsc cpuid tsc_known_freq pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single pti ssbd ibrs ibpb stibp fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm mpx avx512f avx512dq rdseed adx smap clflushopt clwb avx512cd avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves arat md_clear arch_capabilities
```

Are you somehow breaking `chr20` into smaller pieces and then joining them back together at the end? I'm not sure how you have things set up on your end. A little more details would help.


We use this to align the reads to a reference:
```bash
time docker run \
-v /data:/data \
quay.io/biocontainers/pbmm2:1.4.0--h56fc30b_0 pbmm2 align \
--sort -j 71 -J 32 \
--preset CCS \
/data/ref.fa /data/read.fastq /data/output.bam
```

I'm not exactly sure if `--preset CCS` translates to your parameters. But that would explain the difference.

**Comment by kishwarshafin** [comment]
@tahashmi can you also confirm the version of `hap.py` you are using?

**Comment by tahashmi** [comment]
Thanks @kishwarshafin for the quick reply.

For test 1., I used the the chr20 complete data by downloading from your repos and ran through bash script.

For 2. test, I have aligned HG002 data with minimap2 and then created 3 chunks of chr20 with `samtools view -@ 24 -h "${DIR}".bam chr20:1-21481388 -o "${DIR}"_chr20:1-21481388.bam`, `samtools view -@ 24 -h "${DIR}".bam chr20:21481389-42962777 -o "${DIR}"_chr20:21481389-42962777.bam` and `samtools view -@ 24 -h "${DIR}".bam chr20:42962778-64444167 -o "${DIR}"_chr20:42962778-64444167.bam`, ran PEPPER/DV on each chunk and finally mergged VCFs at the end. 

I just re-run exactly the same [PEPPER PacBio](https://github.com/kishwarshafin/pepper/blob/r0.4/docs/pipeline_singularity/HiFi_variant_calling_singularity.md) use case through bash script getting following outputs (using `hap.py_v0.3.12.sif` for benchmarking):

```
  Type Filter  TRUTH.TOTAL  TRUTH.TP  TRUTH.FN  QUERY.TOTAL  QUERY.FP  QUERY.UNK  FP.gt  FP.al  METRIC.Recall  METRIC.Precision  METRIC.Frac_NA  METRIC.F1_Score  TRUTH.TOTAL.TiTv_ratio  QUERY.TOTAL.TiTv_ratio  TRUTH.TOTAL.het_hom_ratio  QUERY.TOTAL.het_hom_ratio
 INDEL    ALL        11256     11168        88        22782       100      11101     56     35       0.992182          0.991439        0.487271         0.991810                     NaN                     NaN                   1.561710                   2.221054
 INDEL   PASS        11256     11168        88        22782       100      11101     56     35       0.992182          0.991439        0.487271         0.991810                     NaN                     NaN                   1.561710                   2.221054
   SNP    ALL        71333     71276        57       102971         9      31607      7      2       0.999201          0.999874        0.306951         0.999537                2.314904                1.879346                   1.715978                   1.943672
   SNP   PASS        71333     71276        57       102971         9      31607      7      2       0.999201          0.999874        0.306951         0.999537                2.314904                1.879346                   1.715978                   1.943672
 
```

My system setup:
```
Architecture:          x86_64
CPU op-mode(s):        32-bit, 64-bit
Byte Order:            Little Endian
CPU(s):                24
On-line CPU(s) list:   0-23
Thread(s) per core:    1
Core(s) per socket:    12
Socket(s):             2
NUMA node(s):          2
Vendor ID:             GenuineIntel
CPU family:            6
Model:                 63
Model name:            Intel(R) Xeon(R) CPU E5-2690 v3 @ 2.60GHz
Stepping:              2
CPU MHz:               2396.875
CPU max MHz:           3500.0000
CPU min MHz:           1200.0000
BogoMIPS:              5188.15
Virtualization:        VT-x
L1d cache:             32K
L1i cache:             32K
L2 cache:              256K
L3 cache:              30720K
NUMA node0 CPU(s):     0-11
NUMA node1 CPU(s):     12-23
Flags:                 fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc aperfmperf eagerfpu pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm epb invpcid_single intel_ppin ssbd ibrs ibpb stibp tpr_shadow vnmi flexpriority ept vpid fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid cqm xsaveopt cqm_llc cqm_occup_llc dtherm ida arat pln pts md_clear spec_ctrl intel_stibp flush_l1d
```
You think does something in results depends on systems resources (like vCPUs or memory)? 

I am bit confused now, Few days back I had run PEPPER ONT test case with `docker://pkrusche/hap.py:latest `  and the results were like this:

 ```
 Type Filter  TRUTH.TOTAL  TRUTH.TP  TRUTH.FN  QUERY.TOTAL  QUERY.FP  QUERY.UNK  FP.gt  METRIC.Recall  METRIC.Precision  METRIC.Frac_NA  METRIC.F1_Score  TRUTH.TOTAL.TiTv_ratio  QUERY.TOTAL.TiTv_ratio  TRUTH.TOTAL.het_hom_ratio  QUERY.TOTAL.het_hom_ratio
 INDEL    ALL        11256      6038      5218        11594       914       4508    329       0.536425          0.871013        0.388822         0.663949                     NaN                     NaN                   1.561710                   1.555780
 INDEL   PASS        11256      6038      5218        11594       914       4508    329       0.536425          0.871013        0.388822         0.663949                     NaN                     NaN                   1.561710                   1.555780
   SNP    ALL        71333     71111       222       110784       249      39399     26       0.996888          0.996512        0.355638         0.996700                2.314904                 1.66885                   1.715978                   1.503188
   SNP   PASS        71333     71111       222       110784       249      39399     26       0.996888          0.996512        0.355638         0.996700                2.314904                 1.66885                   1.715978                   1.503188
```

Now with `hap.py_v0.3.12.sif `the output results are different and still din't match with yours:
 
```
  Type Filter  TRUTH.TOTAL  TRUTH.TP  TRUTH.FN  QUERY.TOTAL  QUERY.FP  QUERY.UNK  FP.gt  FP.al  METRIC.Recall  METRIC.Precision  METRIC.Frac_NA  METRIC.F1_Score  TRUTH.TOTAL.TiTv_ratio  QUERY.TOTAL.TiTv_ratio  TRUTH.TOTAL.het_hom_ratio  QUERY.TOTAL.het_hom_ratio
 INDEL    ALL        11256      6038      5218        11594       918       4508    331    370       0.536425          0.870449        0.388822         0.663784                     NaN                     NaN                   1.561710                   1.555780
 INDEL   PASS        11256      6038      5218        11594       918       4508    331    370       0.536425          0.870449        0.388822         0.663784                     NaN                     NaN                   1.561710                   1.555780
   SNP    ALL        71333     71111       222       110784       249      39399     26    164       0.996888          0.996512        0.355638         0.996700                2.314904                 1.66885                   1.715978                   1.503188
   SNP   PASS        71333     71111       222       110784       249      39399     26    164       0.996888          0.996512        0.355638         0.996700                2.314904                 1.66885                   1.715978                   1.503188

```

**Comment by kishwarshafin** [comment]
@tahashmi ,
For use-case 1: can you please share your script so I can run it locally to see exactly what is happening?

For use case 2:
You should not chunk the genome into such small pieces. Margin is phasing the genome and with such short segments, it's expected the phasing to fail. This is a haplotype-aware pipeline, so ideally you want to leverage the resolution of long reads as much as possible. 

If you want to parallelize the whole genome, you can run each chromosome in parallel. But, chunking the chromosome into smaller pieces is not an ideal use-case. You can possibly chunk by p arm and q arm but anything below that would significantly hurt the accuracy of this pipeline. 

You can read details about the method in this [pre-print](https://www.biorxiv.org/content/10.1101/2021.03.04.433952v1)

**Comment by tahashmi** [comment]
For use case 1, Please check the script and logs in these attachments. 
[pacbio_dv_pepper_log.txt](https://github.com/kishwarshafin/pepper/files/6202252/pacbio_dv_pepper_log.txt)
[run.txt](https://github.com/kishwarshafin/pepper/files/6202253/run.txt)

**Comment by tahashmi** [comment]
Could you please explain bit more how to chunk chromosomes by p arm and q arm?

**Comment by kishwarshafin** [comment]
@tahashmi ,

Update from our side. We were able to reproduce the issue for case 1 you are seeing on your end. The difference between your run and our run is the number of threads. When you are using 16 threads, Margin stitches the genome differently. The issue is from nondeterminism in Margin's stitching and we are fixing it on our end. Thank you very much for reporting it, I will provide an update here very soon.

**Comment by tahashmi** [comment]
Thanks @kishwarshafin  for the updates.

**Comment by kishwarshafin** [comment]
@tahashmi ,

We have pushed a new version out `r0.4.1` in docker, it only changes the version of Margin:

If you only change the commands like this:
```bash
singularity pull docker://kishwars/pepper_deepvariant:r0.4.1

singularity exec --bind /usr/lib/locale/ \
pepper_deepvariant_r0.4.1.sif \
run_pepper_margin_deepvariant call_variant \
-b "${INPUT_DIR}/${BAM}" \
-f "${INPUT_DIR}/${REF}" \
-o "${OUTPUT_DIR}" \
-t ${THREADS} \
--ccs
```
You will get values exactly like this one with `16` or `32` threads:
```bash
 Type  TRUTH.TOTAL  TRUTH.TP  TRUTH.FN  QUERY.FP  FP.gt  FP.al  METRIC.Recall  METRIC.Precision  METRIC.F1_Score
INDEL        11256     11171        85        95     52     33       0.992448          0.991866         0.992157
  SNP        71333     71279        54         7      6      1       0.999243          0.999902         0.999572
```
As this changes the numbers we have uploaded, I will not make it official until the next release where it should be fixed. Thank you again for reporting this. Let me know if you have issues re-generating the numbers I reported here.

**Comment by tahashmi** [comment]
Thanks @kishwarshafin 
I will test it and will let you know about the results.

**Comment by tahashmi** [comment]
Got exactly the same results for PACBIO. 
Also tested `r0.4.1` for ONT and results are still different. Do I expect another another rev for ONT as well?

**Comment by kishwarshafin** [comment]
@tahashmi , 

If you are comparing ONT numbers with the numbers we provided in the readme then it will be different. But if you do two runs, one with 16 threads and one with 32 threads, those two numbers will be the same. So the reproducibility issue has been fixed by adding determinism in the margin's stitching. 

Also yes, we are trying to improve both runtime and accuracy as part of the development process, so we will do some revisions.
